mss/L2Cache.cpp

```diff
@@ -0,0 +1,657 @@
+
+        in_biu_ack_.registerConsumerHandler
+            (CREATE_SPARTA_HANDLER_WITH_DATA(L2Cache, getAckFromBIU_, bool));
+
+        // Pipeline collection config
+        l2cache_pipeline_.enableCollection(node);
+        // Allow the pipeline to create events and schedule work
+        l2cache_pipeline_.performOwnUpdates();
+
+        // There can be situations where NOTHING is going on in the
+        // simulator but forward progression of the pipeline elements.
+        // In this case, the internal event for the pipeline will
+        // be the only event keeping simulation alive.  Sparta
+        // supports identifying non-essential events (by calling
+        // setContinuing to false on any event).
+        l2cache_pipeline_.setContinuing(true);
+
+        l2cache_pipeline_.registerHandlerAtStage(stages_.CACHE_LOOKUP, CREATE_SPARTA_HANDLER(L2Cache, handleCacheAccessRequest_));
+
+        l2cache_pipeline_.registerHandlerAtStage(stages_.HIT_MISS_HANDLING, CREATE_SPARTA_HANDLER(L2Cache, handleCacheAccessResult_));
+
+        // L2 cache config
+        const uint32_t l2_size_kb = p->l2_size_kb;
+        const uint32_t l2_associativity = p->l2_associativity;
+        std::unique_ptr<sparta::cache::ReplacementIF> repl(new sparta::cache::TreePLRUReplacement
+                                                         (l2_associativity));
+        l2_cache_.reset(new olympia::SimpleDL1( getContainer(), l2_size_kb, l2_lineSize_, *repl));
+        
+        ILOG("L2Cache construct: #" << node->getGroupIdx());
+        ILOG("Starting BIU credits = " << l2cache_biu_credits_);
+    }
+
+
+    ////////////////////////////////////////////////////////////////////////////////
+    // Callbacks
+    ////////////////////////////////////////////////////////////////////////////////
+
+    // Receive new L2Cache request from DCache
+    void L2Cache::getReqFromDCache_(const olympia::InstPtr & inst_ptr) {
+        
+        ILOG("Request received from DCache on the port");
+
+        appendDCacheReqQueue_(inst_ptr);
+
+        // Schedule L2Cache request handling event only when:
+        // Request queue is not empty
+        if (dcache_req_queue_.size() < dcache_req_queue_size_) {
+            ev_handle_dcache_l2cache_req_.schedule(sparta::Clock::Cycle(0));
+            num_reqs_from_dcache_++;
+        }
+        else {
+            ILOG("This request cannot be serviced right now, L2Cache input buffer from DCache is already full!");
+        }
+    }
+
+    // Receive new L2Cache request from IL1
+    void L2Cache::getReqFromIL1_(const olympia::InstPtr & inst_ptr) {
+        
+        ILOG("Request received from IL1 on the port");
+
+        appendIL1ReqQueue_(inst_ptr);
+
+        // Schedule L2Cache request handling event only when:
+        // (1)Request queue is not empty
+        if (il1_req_queue_.size() < il1_req_queue_size_) {
+            ev_handle_il1_l2cache_req_.schedule(sparta::Clock::Cycle(0));
+            num_reqs_from_il1_++;
```

Comment:
Another small nit:
```suggestion
            ++num_reqs_from_il1_;
```
The original code will return a _copy_ of the counter after the increment, which (mostly in debug) this is inefficient.  Most of the times the compiler will optimize this out, but don't rely on that.  😉